pkg/engine/events.go

```diff
@@ -28,14 +28,46 @@ import (
 	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/tokens"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/deepcopy"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
 )
 
 // Event represents an event generated by the engine during an operation. The underlying
 // type for the `Payload` field will differ depending on the value of the `Type` field
 type Event struct {
 	Type    EventType
-	Payload interface{}
+	payload interface{}
+}
+
+func NewEvent(typ EventType, payload interface{}) Event {
+	ok := false
+	switch typ {
+	case CancelEvent:
+		ok = payload == nil
+	case StdoutColorEvent:
+		_, ok = payload.(StdoutEventPayload)
+	case DiagEvent:
+		_, ok = payload.(DiagEventPayload)
+	case PreludeEvent:
+		_, ok = payload.(PreludeEventPayload)
+	case SummaryEvent:
+		_, ok = payload.(SummaryEventPayload)
+	case ResourcePreEvent:
+		_, ok = payload.(ResourcePreEventPayload)
+	case ResourceOutputsEvent:
+		_, ok = payload.(ResourceOutputsEventPayload)
+	case ResourceOperationFailed:
+		_, ok = payload.(ResourceOperationFailedPayload)
+	case PolicyViolationEvent:
+		_, ok = payload.(PolicyViolationEventPayload)
+	default:
+		contract.Failf("unknown event type %v", typ)
+	}
+	contract.Assertf(ok, "invalid payload of type %T for event type %v", payload, typ)
```

Comment:
This is pretty slick. TBH I wouldn't have considered adding the check for the type of the payload too. 👍